rfd/0144-client-tools-auto-updates.md

```diff
@@ -0,0 +1,82 @@
+---
+authors: Bernard Kim (bernard@goteleport.com)
+state: draft
+---
+
+# RFD 0144 - Client tools auto update
+
+## Required Approvers
+
+* Engineering: @r0mant && @jimbishopp
+* Product: @klizhentas || @xinding33
+* Security: @jentfoo
+
+## What
+
+A workflow for Teleport client tools (both tsh and tctl) to automatically stay up to date with the latest cloud-stable version of Teleport.
+
+## Why
+
+One of our goals for Teleport Cloud is to increase the cadence at which clusters can be updated.
+This requires that Teleport Agents and Teleport client tools be up to date with the latest compatible version of the cluster.
+
+Manually updating Teleport Agents and client tools can be a tedious task, so customers can end up delaying their updates.
+Simplifying this process will help customers update and stay up to date with their Teleport version.
+
+Development for Automatic Updates of Teleport Agents has already been completed, so the success criteria for this RFD only includes developing a simpler way to keep Teleport client tools up to date.
+
+## Details
+
+Teleport will now store a single cached version of the client tools in the user's $HOME directory. tsh/tctl will execute this cached version when communicating with the Teleport Cloud cluster.
+
+Cloud will host an endpoint for Teleport client tools version discovery. tsh/tctl will check this endpoint to identify the current cloud-stable version of Teleport. If the cached version is behind the current cloud-stable version, the cache will be updated to match the current cloud-stable version. This will ensure Teleport client tools are always executed using the latest compatible version of Teleport when communicating with a Teleport Cloud cluster.
+
+For now this feature will be disabled for self-hosted users and will only be supported for Cloud users.
```

Comment:
How will the client determine whether it's talking to Cloud or self-hosted cluster (i.e. whether to check for update or not)?

Reply:
Hmm, what about including this information within the ping response? Maybe just a `cloud` boolean flag?

Reply:
In this case, can we get the version from the ping itself? The rationale behind hosting a version endpoint separate from Teleport was to recover agent even in case of large breaking changes.

Here if we already have a dependency on teleport, we can pull the version from teleport itself. This will allow easier support when people will ask for self-hosted and mitigates version freeze attacks (someone, malicious or just a zealous sysadmin, would block traffic to the version discovery endopoint).

Reply:
Yeah, we could also get the version information from the ping response. I'll add this as an alternative solution.

The benefit of using a separate version endpoint is that we can upgrade/downgrade the client tools version separately from the cluster version.

Also it sounds like we're leaning towards the idea of removing the minor version information from the ping response.